src/libraries/System.Private.CoreLib/src/System/Buffers/Binary/BinaryPrimitives.ReverseEndianness.cs

```diff
@@ -146,36 +145,38 @@ public static partial class BinaryPrimitives
                 // to end of the source and not have to worry about writing into the destination and clobbering
                 // source data we haven't yet read.
 
-                if (Vector128.IsHardwareAccelerated && source.Length >= Vector128<T>.Count)
+                int i = 0;
+
+                if (Vector256.IsHardwareAccelerated)
                 {
-                    if (Vector256.IsHardwareAccelerated && source.Length >= Vector256<T>.Count)
+                    while (i <= source.Length - Vector256<T>.Count)
                     {
-                        ref T end256 = ref Unsafe.Add(ref sourceRef, source.Length - Vector256<T>.Count + 1);
-                        do
-                        {
-                            Vector256.StoreUnsafe(TReverser.Reverse(Vector256.LoadUnsafe(ref sourceRef)), ref destRef);
-                            sourceRef = ref Unsafe.Add(ref sourceRef, Vector256<T>.Count);
-                            destRef = ref Unsafe.Add(ref destRef, Vector256<T>.Count);
-                        }
-                        while (Unsafe.IsAddressLessThan(ref sourceRef, ref end256));
+                        Vector256.StoreUnsafe(TReverser.Reverse(Vector256.LoadUnsafe(ref sourceRef, (nuint)i)), ref destRef, (nuint)i);
+                        i += Vector256<T>.Count;
                     }
-                    else
+
+                    if (Vector128.IsHardwareAccelerated)
```

Comment:
I think if `Vector256.IsHardwareAccelerated == true` this implies `Vector128.IsHardwareAccelerated == true`, so this check here can be avoided.

At least it was true for AVX -> SSE, and I can't imagine that changed.

Besides that I'd write it as suggested, which is also less code (C# and IL). But leaving it up to you 😉.

Reply:
> Besides that I'd write it as suggested

As suggested where?

Reply:
https://github.com/dotnet/runtime/pull/76339#discussion_r983332324

Reply:
Why is that better? This avoids an extra loop in the 256 case for processing a stray 128 vector.

Reply:
At machine-code level it's just the additional jump back to loop-head (and fail then) which shouldn't cost much.
But it eleminates the (C#) code-duplication.

Although the loop bodies are quite small...so I have no strong preference here. Except the Vec25 -> Vec128 supported checks.

Reply:
> I think if Vector256.IsHardwareAccelerated == true this implies Vector128.IsHardwareAccelerated == true, so this check here can be avoided.

There is no guarantee of that and it is the case that while `Vector128.IsHardwareAccelerated` might be true, `Vector64.IsHardwareAccelerated` could be false.

Platforms that only have variable width vector support might fall into this more general camp (RISC-V for example has both a packed-simd and a variable-simd ISA, neither is required and a chipset could provide "just one").

These are constant folded by the JIT so its fine to do the more verbose/proper check.